jaraco/compat/py38.py

```diff
@@ -12,33 +12,43 @@
 import functools
 import sys
 import types
+from typing import Protocol, cast
 
 
-def _fixer(orig: str):  # pragma: no cover
+class _RFixed(Protocol):
+    def removesuffix(self, suffix: str) -> str: ...
+    def removeprefix(self, prefix: str) -> str: ...
+
+
+def _fixer(orig: str) -> _RFixed:  # pragma: no cover
     """
     Return an object that implements removesuffix and removeprefix on orig.
     """
 
-    def removesuffix(suffix):
+    def removesuffix(suffix: str) -> str:
         # suffix='' should not call orig[:-0].
         if suffix and orig.endswith(suffix):
             return orig[: -len(suffix)]
         else:
             return orig[:]
 
-    def removeprefix(prefix):
+    def removeprefix(prefix: str) -> str:
         if orig.startswith(prefix):
             return orig[len(prefix) :]
         else:
             return orig[:]
 
-    return types.SimpleNamespace(removesuffix=removesuffix, removeprefix=removeprefix)
+    return cast(
+        _RFixed,
+        types.SimpleNamespace(removesuffix=removesuffix, removeprefix=removeprefix),
+    )
 
 
-r_fix = _fixer if sys.version_info < (3, 9) else lambda x: x
+if sys.version_info < (3, 9):  # pragma: no cover
+    cache = functools.lru_cache(maxsize=None)
+    r_fix = _fixer
+else:
+    cache = functools.cache
 
-cache = (
-    functools.cache  # type: ignore[attr-defined]
-    if sys.version_info >= (3, 9)
-    else functools.lru_cache(maxsize=None)
-)
+    def r_fix(orig: str) -> str:
+        return orig
```

Comment:
I do prefer to have single assignment rather than conditional assignment. I also like to keep concerns separate. With this change, now r_fix and cache is interleaved.

But I see now why you do it - mypy doesn't honor conditionals that aren't conditional blocks.

The problem with accepting this change is I'm already tempted to refactor it to separate the concerns and consolidate assignments.

I think this is a case where the value of the original logic is greater than the value of the type checking.

Restored in 67d3f31.

Reply:
It's definitely a balance of what you value!
Pyright understands "ternary" (conditional expression), by not mypy https://github.com/python/mypy/issues?q=sort%3Aupdated-desc+is%3Aopen+label%3Atopic-ternary-expression (the most specific issue/request related to this case would be https://github.com/python/mypy/issues/13003 )

Generally as long as the exposed type is correct.